res/controllers/lib/mididispatcher.mjs

```diff
@@ -0,0 +1,26 @@
+export class MidiDispatcher {
+    constructor(noteOff) {
+        this.noteOff = noteOff;
+        this.inputMap = new Map();
```

Comment:
It’s not clear whether direct access to inputMap is allowed or not. If not, how would unbinding callbacks be handled?

Reply:
inputMap is not meant to be manipulated directly... too bad JS has no private variables.

Callbacks can be replaced by simply registering a new callback with the same midiBytes. I can't think of a use case for unregistering a callback without replacing it, but if you do want to do this, you could pass `null` for the callback.

Reply:
There are private members in ES2019. For now we could use the underscore convention and prefix private members with an underscore, but IMHO this is unnecessary. Instead, we should add API docs to the controller mapping section of the manual and consider everything not documented there as an implementation detail that might change at any time.

Reply:
Considering that it's not possible to register multiple callbacks for the same midi bytes, maybe we should make this more explicit by naming the method `setInputCallback`?

Reply:
Thinking about this more, I feel like we’re making the same mistake as with the CO callbacks. Maybe it would make more sense if we changed the design to something where you get a handle for your callback so you can have multiple handlers (possibly allowing for concurrent execution) for the same input (also allowing them being disconnected via their handle).

Reply:
Interesting idea. What would be a use case for connecting multiple callbacks for one incoming MIDI message? I don't think I have ever wanted to do this. Would you ever have two Components responding to the same MIDI message? I don't think so.

On the other hand, having similar APIs for input and output could be nice.

Reply:
Forgot about this.

> Would you ever have two Components responding to the same incoming MIDI message?

It depends. I can't think of an example right now, but there is the hypothetical case where a single midi message might contain several meanings in which case I'd create several components to separate their responsibility.

Reply:
I'm fairly confident that never happens with MIDI messages, at least the regular 3 byte messages this class handles. Maybe it could be appropriate for sysex messages, but this class does not handle those.

Reply:
Yeah, in case it does happen, one could still hack around it by supplying a "proxy" handler that calls all the other handlers.